File: DiagnolTraverse.py
#Time Complexity: O(mn)
#Space Complexity : O(mn)

class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        row = len(mat)
        col = len(mat[0])

        res = []
        cur_row = cur_col = 0
        going_up = True

        while len(res) != row * col:
            if going_up:
                while cur_row >= 0 and cur_col < col:
                    res.append(mat[cur_row][cur_col])

                    cur_row -= 1
                    cur_col += 1

                if cur_col == col:
                    cur_row += 2
                    cur_col -= 1
                else:
                    cur_row += 1
                going_up = False 
            
            else:
                while cur_col >= 0 and cur_row < row:
                    res.append(mat[cur_row][cur_col])

                    cur_row += 1
                    cur_col -= 1

                if cur_row == row:
                    cur_col += 2
                    cur_row -= 1
                else:
                    cur_col += 1
                going_up = True
        
        return res

File: DiagnolTraverse.java
// TC: O(m*n) m = no. of rows, n=no. of columns
// SC: O(1) No auxiliary space is used
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class DiagnolTraverse {
    public int[] findDiagonalOrder(int[][] mat) {
        int m = mat.length;
        int n = mat[0].length;
        int result[] = new int[m*n];
        int row = 0;
        int col = 0;
        boolean direction = true;
        for(int i = 0;i<result.length;i++) {
            result[i] = mat[row][col];
            if(direction) {
                if(row==0 && col!= n-1) {
                    col ++;
                    direction = false;
                } else if(col == n-1) {
                    row++;
                    direction = false;
                } else{
                    row -- ; col++;
                }
            } else{
                if(col==0 && row!=m-1) {
                    row++;
                    direction = true;
                } else if(row==m-1) {
                    col ++;
                    direction = true;
                } else {
                    row++; col--;
                }
            }
        }
        return result;
    }
}